### PR TITLE
Prevent RazorErrorExtensions.ToDiagnostics from throwing when it encounters SourceLocation.Undefined \ negative error lengths

### PR DESCRIPTION
**Pull request by pranavkm**

### CODE REVIEW

**Comment by pranavkm** [comment]
Old PR: https://github.com/aspnet/Mvc/pull/2257

**Comment by Eilon** [comment]
:shipit:

File: src/Microsoft.AspNet.Mvc.Razor/Precompilation/RazorErrorExtensions.cs [hunk 1]
```diff
@@ -20,15 +22,19 @@ public static Diagnostic ToDiagnostics([NotNull] this RazorError error, [NotNull
...
-                                                   error.Location.CharacterIndex + error.Length);
-            var linePositionSpan = new LinePositionSpan(linePositionStart, linePositionEnd);
-
-            var location = Location.Create(filePath, textSpan, linePositionSpan);
+            var location = error.Location;
+            if (location == SourceLocation.Undefined)
+            {
+                location = SourceLocation.Zero;
+            }
+            var length = Math.Max(0, error.Length);
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor/Precompilation/RazorErrorExtensions.cs]: suggest using a fixed and short length when `location == SourceLocation.Zero`.  end up highlighting a random amount of text in the current (unrelated) file as-is.

File: src/Microsoft.AspNet.Mvc.Razor/Precompilation/RazorErrorExtensions.cs [hunk 2]
```diff
@@ -20,15 +22,19 @@ public static Diagnostic ToDiagnostics([NotNull] this RazorError error, [NotNull
...
 
-            return Diagnostic.Create(descriptor, location);
+            var textSpan = new TextSpan(location.AbsoluteIndex, length);
+            var linePositionStart = new LinePosition(location.LineIndex, location.CharacterIndex);
+            var linePositionEnd = new LinePosition(location.LineIndex,
+                                                   location.CharacterIndex + length);
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor/Precompilation/RazorErrorExtensions.cs]: nit: can this fit on one line? if not wrap first argument as well.

**Comment by dougbu** [comment]
:shipit: w/ one suggested change

File: src/Microsoft.AspNet.Mvc.Razor/Precompilation/RazorErrorExtensions.cs (hunk 1 above)
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/Precompilation/RazorErrorExtensions.cs]: The editor ends up placing the cursor at the start of the error column when we do set the length to 0. Seems like an ok-ish behavior - so I'm leaving this as is.

File: src/Microsoft.AspNet.Mvc.Razor/Precompilation/RazorErrorExtensions.cs (hunk 2 above)
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/Precompilation/RazorErrorExtensions.cs]: :ok:

File: src/Microsoft.AspNet.Mvc.Razor/Precompilation/RazorErrorExtensions.cs (hunk 1 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor/Precompilation/RazorErrorExtensions.cs]: I was talking about the case where `error.Length > 0` and `location == SourceLocation.Zero` (might have been `SourceLocation.Undefined` a few lines earlier).  I would prefer `length == 0` for this case, e.g.
``` c#
var location = error.Location;
if (location == SourceLocation.Undefined)
{
  location = SourceLocation.Zero;
}

var length = Math.Max(0, error.Length);
if (location == SourceLocation.Zero)
{
  length = 0;
}
```

**Pull request closed without merging**